packages/vite/src/node/server/index.ts

```diff
@@ -583,84 +583,73 @@ async function startServer(
   }
 
   const options = server.config.server
-  let port = inlinePort || options.port || 3000
+  const port = inlinePort || options.port || 3000
   const hostname = resolveHostname(options.host)
 
   const protocol = options.https ? 'https' : 'http'
   const info = server.config.logger.info
   const base = server.config.base
 
   return new Promise((resolve, reject) => {
-    const onError = (e: Error & { code?: string }) => {
-      if (e.code === 'EADDRINUSE') {
-        if (options.strictPort) {
-          httpServer.removeListener('error', onError)
-          reject(new Error(`Port ${port} is already in use`))
-        } else {
-          info(`Port ${port} is in use, trying another one...`)
-          httpServer.listen(++port, hostname.host)
-        }
-      } else {
-        httpServer.removeListener('error', onError)
-        reject(e)
-      }
-    }
-
-    httpServer.on('error', onError)
-
-    httpServer.listen(port, hostname.host, () => {
-      httpServer.removeListener('error', onError)
-
-      info(
-        chalk.cyan(`\n  vite v${require('vite/package.json').version}`) +
-          chalk.green(` dev server running at:\n`),
-        {
-          clear: !server.config.logger.hasWarned
-        }
-      )
-
-      printServerUrls(hostname, protocol, port, base, info)
-
-      // @ts-ignore
-      if (global.__vite_start_time) {
+    httpServerStart(httpServer, {
+      port,
+      strict: options.strictPort,
+      host: hostname.host,
+      logger: server.config.logger
+    })
+      .then((serverPort) => {
```

Comment:
Also here :thinking: 
We have an async function returning a Promise and that calls `.then` :eyes: 
Can you try to just use async-await? :slightly_smiling_face:

Reply:
I have thought about whether to modify it or not, but I need to keep `catch` here, so I think the modification is meaningless.

Reply:
And `dev server` must return `Promise<ViteDevServer>`
https://github.com/vitejs/vite/blob/61dbb3a91b1e9d05944bbd7c5f0e23e4c5f6a4c7/packages/vite/src/node/server/index.ts#L575-L579

use `async-await`, then the callback function needs to add `async`
```javascript
return new Promise(async (resolve, reject) => {}
```

this is not allowed
![image](https://user-images.githubusercontent.com/7699524/125641199-3f73cad4-9785-41e2-b22b-4267a19d3546.png)

Reply:
Sorry, maybe I didn't understand your intentions.
If you can, please use some code to illustrate it. Maybe I will understand.

Reply:
Cannot submit a suggestion 🙁 here is the whole function:

```ts
async function startServer(
  server: ViteDevServer,
  inlinePort?: number,
  isRestart: boolean = false
): Promise<ViteDevServer> {
  const httpServer = server.httpServer
  if (!httpServer) {
    throw new Error('Cannot call server.listen in middleware mode.')
  }

  const options = server.config.server
  const port = inlinePort || options.port || 3000
  const hostname = resolveHostname(options.host)

  const protocol = options.https ? 'https' : 'http'
  const info = server.config.logger.info
  const base = server.config.base

  const serverPort = await httpServerStart(httpServer, {
    port,
    strict: options.strictPort,
    host: hostname.host,
    logger: server.config.logger
  })

  info(
    chalk.cyan(`\n  vite v${require('vite/package.json').version}`) +
      chalk.green(` dev server running at:\n`),
    {
      clear: !server.config.logger.hasWarned
    }
  )

  printServerUrls(hostname, protocol, serverPort, base, info)

  // @ts-ignore
  if (global.__vite_start_time) {
    info(
      chalk.cyan(
        // @ts-ignore
        `\n  ready in ${Date.now() - global.__vite_start_time}ms.\n`
      )
    )
  }

  // @ts-ignore
  const profileSession = global.__vite_profile_session
  if (profileSession) {
    profileSession.post('Profiler.stop', (err: any, { profile }: any) => {
      // Write profile to disk, upload, etc.
      if (!err) {
        const outPath = path.resolve('./vite-profile.cpuprofile')
        fs.writeFileSync(outPath, JSON.stringify(profile))
        info(
          chalk.yellow(`  CPU profile written to ${chalk.white.dim(outPath)}\n`)
        )
      } else {
        throw err
      }
    })
  }

  if (options.open && !isRestart) {
    const path = typeof options.open === 'string' ? options.open : base
    openBrowser(
      `${protocol}://${hostname.name}:${serverPort}${path}`,
      true,
      server.config.logger
    )
  }

  return server
}
```

Reply:
Thank you for your guidance, it has been submitted 😄